torch/testing/_internal/common_methods_invocations.py

```diff
@@ -4,14 +4,133 @@
+                 *,
+                 dtypes=None,  # dtypes this function is expected to work with
+                 dtypesIfCPU=None,  # dtypes this function is expected to work with on CPU
+                 dtypesIfCUDA=None,  # dtypes this function is expected to work with on CUDA
+                 dtypesIfROCM=None,  # dtypes this function is expected to work with on ROCM
+                 has_out_kwarg=None,  # whether this function has an out kwarg
+                 decorators=None,  # decorators to apply to generated tests
+                 tests_to_skip=None):  # regular expressions describing tests to skip
+        self.name = name
+
+        self.dtypes = dtypes
+        self.dtypesIfCPU = dtypesIfCPU
+        self.dtypesIfCUDA = dtypesIfCUDA
+        self.dtypesIfROCM = dtypesIfROCM
+
+        self.method_variant = getattr(torch.Tensor, name) if hasattr(torch.Tensor, name) else None
+        inplace_name = name + "_"
+        self.inplace_variant = getattr(torch.Tensor, inplace_name) if hasattr(torch.Tensor, name) else None
+        self.has_out_kwarg = has_out_kwarg
+
+        self.decorators = decorators
+        self.tests_to_skip = tests_to_skip
+
+    # Returns the op (currently only supports torch.<op_name>)
+    def getOp(self):
+        return getattr(torch, self.name)
+
+    # Returns the method variant of the op (torch.Tensor.<op_name>)
+    # or None if the op has no method variant
+    def getMethod(self):
+        return self.method_variant
+
+    # Returns the inplace variant of the op (torch.Tensor.<op_name>_)
+    # or None if the op has no inplace variant
+    def getInplace(self):
+        return self.inplace_variant
+
+    # Returns True if the test should be skipped and False otherwise
+    def shouldSkip(self, test_name, *, device, dtype):
+        if self.tests_to_skip is None:
+            return False
+
+        if device is None and dtype is None:
+            full_name = test_name
+        elif dtype is None:
+            assert device is not None
+            full_name = test_name + "_" + str(torch.device(device).type)
+        else:
+            assert device is not None
+            assert dtype is not None
+            full_name = test_name + "_" + str(torch.device(device).type) + "_" + str(dtype).split('.')[1]
+
+        for regex in self.tests_to_skip:
+            if re.match(regex, full_name):
+                return True
+
+        return False
+
+
+# Metadata class for unary "universal functions (ufuncs)" that accept a single
+# tensor and have common properties like:
+#   - they are elementwise functions
+#   - the input shape is the output shape
+#   - they typically have method and inplace variants
+#   - they typically support the out kwarg
+#   - they typically have NumPy or SciPy references
+# See NumPy's universal function documentation
+# (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+# about the concept of ufuncs.
+class UnaryUfuncMeta(OpMeta):
```

Comment:
Is subclassing the right overriding mechanism here? I'm kind of imagining that there might be various ways to describe what functions are (e.g., a function that is "linear", or "needs to work with 64-bit") and if you have a lot of descriptions like this, a subclass hierarchy doesn't work too good for it. Better to have something more unstructured like tags or something.

It doesn't seem like you're using any of the extra juice that OO gives you here; you could avoid a future refactor by making the "named tuples" here (as you have described them) even more dumb.

Reply:
The benefit of subclassing here is setting the default values in the constructor. We could change the system to be tag-based and look up the default values based on tags?

Reply:
Since this can be changed easily in the future, I suggest we table it for now.